moler/connection_observer.py

```diff
@@ -19,10 +19,12 @@
 from moler.helpers import camel_case_to_lower_case_underscore
 from moler.helpers import instance_id
 from moler.runner import ThreadPoolExecutorRunner
+import time
 
 
 @add_metaclass(ABCMeta)
 class ConnectionObserver(object):
+    _list_of_exceptions = list()  # list of dict: "exception" and "time"
```

Comment:
it is shared resource for all observers. These observers may run in different threads, so:

1. we need lock to protect it: `self._exceptions_lock = threading.Lock()`
1. all modifications of shared resource should be lock-protected

I would change name of shared-resource into `_not_reraised_exceptions`

Reply:
So far we don't delete items from the list so it is thread safe, however if we want to remove lock must be implemented.

Reply:
No, it's not. `get_active_exceptions_in_time()` just reads from that shared resource. What if there is a race: connection observer thread is just setting exception, threads-switch: Python is in main thread just finishing keyword - it calls `_final_check()` which calls `get_active_exceptions_in_time()` and there is no exception collected.

If there would be lock:
```
    def set_exception(self, exception):
        """Should be used to indicate some failure during observation"""
        with self._exceptions_lock:
            self._is_done = True
            self._exception = exception
            self._exception_time = time.time()
            self._needed_exception_raise = True
            ConnectionObserver._append_active_exception(exception, self._exception_time)
```
then process of setting exception is like database-atomic-action "all or none". So, if it was started we are sure it will be collected by .`get_active_exceptions_in_time()`.

Reply:
Doesn't matter. Implementation is changed to remove exceptions so lock is necessary..,

Reply:
Lock is used and exceptions are removed from list.